exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -307,3 +315,235 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+	log := ctrl.LoggerFrom(ctx)
+
+	var noKind bool
+	var noSelector bool
+
+	var infraMachineKind string
+	if err := util.UnstructuredUnmarshalField(infraMachinePool, &infraMachineKind, "status", "infrastructureMachineKind"); err != nil {
+		if errors.Is(err, util.ErrUnstructuredFieldNotFound) {
+			noKind = true
+		} else {
+			return errors.Wrapf(err, "failed to retrieve infraMachineKind from infrastructure provider for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+		}
+	}
+
+	var infraMachineSelector metav1.LabelSelector
+	if err := util.UnstructuredUnmarshalField(infraMachinePool, &infraMachineSelector, "status", "infrastructureMachineSelector"); err != nil {
+		if errors.Is(err, util.ErrUnstructuredFieldNotFound) {
+			noSelector = true
+		} else {
+			return errors.Wrapf(err, "failed to retrieve infraMachineSelector from infrastructure provider for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+		}
+	}
+
+	// If one missing but not both, return error
+	if noKind != noSelector {
+		return errors.Errorf("only one of infraMachineKind or infraMachineSelector found, both must be present")
+	}
+
+	// If MachinePool Machines aren't supported by the InfraMachinePool, skip reconcile and return early.
+	if noKind && noSelector {
+		log.V(2).Info("MachinePool Machines not supported, no infraMachineKind or infraMachineSelector found")
+		return nil
+	}
+
+	log.Info("Reconciling MachinePool Machines", "infrastructureMachineKind", infraMachineKind, "infrastructureMachineSelector", infraMachineSelector)
+	var infraMachineList unstructured.UnstructuredList
+
+	infraMachineList.SetAPIVersion(infraMachinePool.GetAPIVersion())
+	infraMachineList.SetKind(infraMachineKind)
+
+	sampleInfraMachine := &unstructured.Unstructured{}
+	sampleInfraMachine.SetAPIVersion(infraMachinePool.GetAPIVersion())
+	sampleInfraMachine.SetKind(infraMachineKind)
+	// Add watcher for infraMachine, if there isn't one already.
+	// Ensure we add a watch to the infraMachine, if there isn't one already.
+	if err := r.externalTracker.Watch(log, sampleInfraMachine, handler.EnqueueRequestsFromMapFunc(infraMachineToMachinePoolMapper)); err != nil {
+		return errors.Wrapf(err, "failed to add watcher on infraMachine %q", sampleInfraMachine.GroupVersionKind())
+	}
+
+	if err := r.Client.List(ctx, &infraMachineList, client.InNamespace(mp.Namespace), client.MatchingLabels(infraMachineSelector.MatchLabels)); err != nil {
+		return errors.Wrapf(err, "failed to list infra machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	machineList := &clusterv1.MachineList{}
+	labels := map[string]string{
+		clusterv1.MachinePoolNameLabel: mp.Name,
+		clusterv1.ClusterNameLabel:     mp.Spec.ClusterName,
+	}
+
+	if err := r.Client.List(ctx, machineList, client.InNamespace(mp.Namespace), client.MatchingLabels(labels)); err != nil {
+		return err
+	}
+
+	if err := r.deleteMachinesWithoutInfrastructure(ctx, mp, machineList.Items, infraMachineList.Items); err != nil {
```

Comment:
Q: Does this mean we have two deletion flows for MachinePool InfraMachines?

The regular one:
* Machine gets deleted => then Machine.reconcileDelete deletes InfraMachine

The one supported here:
* InfraMachine gets deleted => then deleteMachinesWithoutInfrastructure deletes Machines

Q: Is it possible to always delete the Machine first to end up with a consistent deletion flow?

This way we don't have to worry about edge cases, e.g. like the one for which we need the change in machine_controller_phases.go. This would ensure we always go through the well-tested regular Machine deletion flow and our regular Machine.reconcile can assume that in the normal non-error case an InfraMachine exists

Reply:
The second case, where the InfraMP deletes the InfraMachines first is the existing use case where the user scales down a MachinePool, the InfraMP scales down its own instances, and the CRDs are updated to reflect that. We consider this to be the regular use case as we consider the `providerIDList`/`InstanceList` to be the source of truth and the list of Machines to reflect that. From my understanding, the reason we support the first use case is for consistency purposes, i.e. if you delete a Machine we'd expect it to do something.

Reply:
This also implies that MachinePool Machines don't go through the regular Machine deletion flow, which e.g. means the nodes are not drained. I thought this was a major motiviation of the proposal.

Reply:
I joined the MPM project after the proposal was merged so I can't say what people thought then. From my understanding, the goal was to create Machines as a way to introspect the instances of a MachinePool since right now we only really have the `providerIDList`. This way we can use cluster autoscaler to add the delete machine label so that MPs work with the autoscaler. But ultimately the InfraMP is responsible for making the deletion/creation happen instead of CAPI.

Reply:
I think it's still fine that the InfraMachinePool controller decides which Machines to create/delete. I'm just saying that the InfraMachinePool controller should delete the Machine and not the InfraMachine.

Otherwise it's impossible to leverage the reconcileDelete logic in the Machine controller for node drain, deletion, considering pre-terminate, pre-drain hooks etc. Because if we delete the InfraMachine the actual infrastructure/server/node is already gone before the Machine controller runs through reconcileDelete.

Reply:
Right now, the MachinePool controller deletes the Machine once it detects the InfraMachine is gone. When that happens, it should still trigger the Machine controller, so I believe we can still configure how the Machine controller handles it. Is that what you want us to do? Tagging @fabriziopandini as well.

Reply:
Ideally i would prefer that the MachinePool controller only deletes the Machine.

The Machine reconcileDelete then triggers InfraMachine deletion which is then a no-op for the AzureMachine controller because it sees that the infrastructure is already gone.

This way we would only have tos upport one Machine deletion workflow.

We have the same discussion in the other thread. Basically I don't understand why if the infrastructure is gone it is necessary to first delete the InfraMachine and then the Machine. Shouldn't we be able to always delete the Machine first?